H-866: Implement pinned entity types on User and Org profile pages

benwerner01 opened the pull request:
## 🌟 What is the purpose of this PR?

<!-- Explain, at a high level, what this does and why. -->
<!-- Use the 'What does this change?' section to list more specific implementation details. -->

This PR adds the ability to pin entity types to a user's profile page.

In addition, it addressed the outstanding review comments left in https://github.com/hashintel/hash/pull/3319.

Remaining todos:

- [x] pluralise entity type tab titles
- [x] use link icon in edit pinned entity types modal

## 🔗 Related links

<!-- Add links to any context it is worth capturing (e.g. Issues, Discussions, Discord) -->
<!-- Mark any links which are not publicly accessible as _(internal)_ -->
<!-- Don't rely on links to explain the PR, especially internal ones: use the sections above -->

- H-866

## 🔍 What does this change?

<!-- Use a bullet list to explain your changes in more detail, if it would be helpful. -->
<!-- If applicable, link to the specific commit.-->

See commits.

## Pre-Merge Checklist 🚀

### 🚢 Has this modified a publishable library?

<!-- Confirm you have taken the necessary action to record a changeset or publish a change, as appropriate -->
<!-- Tick AT LEAST ONE box and delete the rest. Do not delete this section! see libs/README.md for info on publishing -->

This PR:

- [x] does not modify any publishable blocks or libraries, or modifications do not need publishing

### 📜 Does this require a change to the docs?

<!-- If this adds a user facing feature or modifies how an existing feature is used, it likely needs a docs change. -->
<!-- Tick ONE box and delete the rest. Do not delete this section! -->

The changes in this PR:

- [x] are internal and do not require a docs change

### 🕸️ Does this require a change to the Turbo Graph?

<!-- If this adds or moves an existing package, modifies `scripts` in a `package.json`, it likely needs a turbo graph change. -->
<!-- Tick ONE box and delete the rest. Do not delete this section! -->

The changes in this PR:

- [x] do not affect the execution graph

## ⚠️ Known issues

<!-- Are there known issues / intentionally omitted functionality? Flag them here to save reviewers doing so -->

## 🐾 Next steps

<!-- Are there are planned/suggested follow ups which are related but won't be done in this PR? -->

## 🛡 What tests cover this?

<!-- What automated tests cover this? Existing ones? New ones? None? -->

None

## ❓ How to test this?

<!-- Tell reviewers how they can test the functionality -->

❗️ This PR cannot be tested in the Vercel deployment because it requires backend changes to be applied

1. Go to your user profile page
2. Toggle the tabs, there should be "Profile" and "Pages" initially
3. Add additional pinned entity type tabs by clicking "Modify"

## 📹 Demo

<!-- Add a screenshot or video showcasing your work -->

https://github.com/hashintel/hash/assets/42802102/c25eefab-757f-485e-9cae-693d3efa6ccc

nonparibus commented:
@benwerner01 

1. The icons for the "TYPES" listed in the "Modify" modal all appear to be those for entity types, even where link types are listed as the featured thing.
1. What happens when you view a page with 0 matches for a pinned type?
1. Please can we change the 'sort by' dropdowns to read `Recency (newest first)` and `Recency (oldest first)`? Is it trivial to add alphanumerical ascending/descending sort in addition? (`Alphabetized (A-Z)` and `Alphabetized (Z-A)` respectively)

benwerner01 commented:
> The icons for the "TYPES" listed in the "Modify" modal all appear to be those for entity types, even where link types are listed as the featured thing.

Will change the icon for link entity types, have captured this in the PR description

> What happens when you view a page with 0 matches for a pinned type?

@nonparibus currently it looks liked this:
![image](https://github.com/hashintel/hash/assets/42802102/d863a2c8-5f69-487d-873b-2e916bd8c97c)

Do you want to add some custom copy when there's no entities to display? And maybe a `Create X` button to create an instance of the type?

> Please can we change the 'sort by' dropdowns to read Last edited (newest first) and Last edited (oldest first)? Is it trivial to add alphanumerical ascending/descending sort (Alphabetized (A-Z) and Alphabetized (Z-A) respectively) in addition, along with Creation date (newest first) and Creation date (oldest first)?

I've modified the copy and added alphabetical sorting in https://github.com/hashintel/hash/pull/3338/commits/d40d59a7b42ccea866403cae2fa21c14ebe355fa

> “Add cover image” text is too large in size on profile page (when in ‘edit’ mode on info). Can we drive-by reduce this to correct it?

Thanks for spotting, have reduced in to `12px` in https://github.com/hashintel/hash/pull/3338/commits/ea6b1c9d0266c08094ca29f6789b5725f179d47d

nonparibus commented:
@benwerner01 

> Do you want to add some custom copy when there's no entities to display? And maybe a Create X button to create an instance of the type?

Yes. Let's also add a loading skeleton while it loads.

Please do the latter while I whip up a design for the former in Figma.

nonparibus commented:
@benwerner01 I've got a duplicate entity type appearing in the list a few times. This may just be a dev/staging bug, but we should probably make sure logic for deduplicating the list is sound, and ensuring that all 5 selected entity types are different. I don't think we're showing already-listed entity types in the selector list, as I wasn't able to add the entity type an additional time, but we should also make sure this is the case. Maybe removal/re-adding causes duplication? I haven't tried to debug why, so will leave to you.

![Capture 2023-10-09 at 11 11 22@2x](https://github.com/hashintel/hash/assets/6226576/d878f4d2-5c03-4f44-b697-0f5ff7237bc5)

CiaranMn commented on apps/hash-frontend/src/pages/[shortname].page.tsx:
```diff
@@ -96,6 +101,215 @@ const ProfilePage: NextPageWithLayout = () => {
+        kind: "profile",
+        title: "Profile",
+      },
+      {
+        kind: "profile-pages",
+        title: "Page",
+        pluralTitle: "Pages",
+      },
```
Could we just add the Page's entityTypeId into `pinnedEntityTypeBaseUrls` automatically here and let the functions to handle other pinned types deal with fetching and sorting it into groups?

We only need to special case it to stop people removing it, the rest of the functionality should be the same, no?

benwerner01 replied:
I can do that, but I'm not really convinced this is worth it. It'd complicate updating `pinnedEntityTypeBaseUrls` in all contexts for something we know we'll always be special casing (when we discussed in standup the other day we concluded we'd want users/orgs to have the non-removable `Pages` tab for the foreseeable future).

It'd also need to be added wherever we create users, so that `pinnedEntityTypeBaseUrls` is populated correctly.

benwerner01 replied:
Ah I get it, apologies for misunderstanding I should've thought of that. Agree this would simplify things, have addressed in https://github.com/hashintel/hash/pull/3338/commits/207953360b4752b0ede415934f7f88c48fad71f3

CiaranMn commented on apps/hash-frontend/src/pages/[shortname].page/edit-pinned-entity-types-modal.tsx:
```diff
@@ -0,0 +1,424 @@
+import { XMarkRegularIcon } from "../../shared/icons/x-mark-regular-icon";
+import { Button, Link } from "../../shared/ui";
+import { entityTypeIcons } from "../../shared/use-entity-icon";
+import { EntityTypeSelector } from "../[shortname]/entities/[entity-uuid].page/create-entity-page/entity-type-selector";
```
This should be in a shared folder

benwerner01 replied:
Addressed in https://github.com/hashintel/hash/pull/3338/commits/579e8cfa9b036947cb67ab568dbb6435b96ee96f

CiaranMn commented on apps/hash-frontend/src/pages/[shortname].page/edit-pinned-entity-types-modal.tsx:
```diff
@@ -0,0 +1,424 @@
+      },
+    });
+
+    /** @todo: error handling */
+
+    setLoading(false);
```
If you called `updateEntity` via `useMutation` rather than the BP hook you could use the loading and error return from the ApolloClient hook rather than manually here. Maybe not worth bothering with since we will remove ApolloClient at some point, but probably not for a while and we will replace it with some similar hook when we do

benwerner01 replied:
Good idea, have done so in https://github.com/hashintel/hash/pull/3338/commits/ebe7262785fe3b48e66d9e4966b52014a2393208

CiaranMn commented on apps/hash-frontend/src/pages/[shortname].page/pinned-entity-type-tab-contents.tsx:
```diff
@@ -0,0 +1,358 @@
+              textTransform: "uppercase",
+            }}
+          >
+            {updatedAtHumanReadable}
```
Could we have a tooltip here with the full date?

benwerner01 replied:
Yes, added in https://github.com/hashintel/hash/pull/3338/commits/470d40000fa0417c0a615e4511664fe2cce7830a

Looks like this:
<img width="530" alt="image" src="https://github.com/hashintel/hash/assets/42802102/e867d7aa-40f5-4ee5-94e3-14343cffc443">

CiaranMn replied:
Looks good, thank you

CiaranMn commented on apps/hash-frontend/src/pages/[shortname].page/edit-pinned-entity-types-modal.tsx:
```diff
@@ -0,0 +1,424 @@
+            Types
+          </Typography>
+          <DragDropContext onDragEnd={handleOnDragEnd}>
+            <Droppable droppableId="pinnedEntityTypes">
```
If possible I think it would be better to lock the drag area to this element so that users can't drag the item all around the screen.

benwerner01 replied:
I spent about 15 minutes trying to get this to work using the `getContainerForClone` prop, and reworking the custom `useDraggableInPortal` hook that I had to add, but couldn't get something working. I think it'd be a good idea, but not sure it's worth investing more time - I can create a follow-up ticket?

CiaranMn replied:
I wouldn't worry about a follow up, we can change it if it becomes annoying.

Out of interest why the new dnd library? `dnd-kit` handles locking the draggable area / axes with a couple of props.

benwerner01 replied:
To be honest mainly because I had used it before, and I thought `@dnd-kit` didn't support the animations when rearranging but having taken another look at the docs I was wrong about that. It may have been a better choice, would you like me to replace it? Or do it in follow-up?

benwerner01 replied:
Have created the ticket: https://linear.app/hash/issue/H-1013/replace-react-beautiful-dnd-with-dnd-kit-in-editpinnedentitytypesmodal

CiaranMn replied:
Thank you

CiaranMn commented on apps/hash-frontend/src/pages/[shortname].page/edit-pinned-entity-types-modal.tsx:
```diff
@@ -0,0 +1,424 @@
+                  }}
+                >
+                  {fields.map((field, index) => (
+                    <Draggable
```
I would disable dragging / hide the drag handle if there is only one item

benwerner01 replied:
Good idea, have done so in https://github.com/hashintel/hash/pull/3338/commits/bbb34228bbac17591f552354bbc1b535511e4215

CiaranMn replied:
Thank you

CiaranMn commented on apps/hash-frontend/src/pages/[shortname].page/profile-page-tabs.tsx:
```diff
@@ -0,0 +1,47 @@
+    currentTab.kind === "profile" ? currentTab.title : currentTab.pluralTitle;
+
+  return (
+    <Box sx={{ overflowX: "scroll" }}>
```
We could made this `auto` so that the scrollbar isn't introduced unless necessary. There will only be layout shift if someone is resizing the window and it appears. We probably also want to style the scrollbar to make it less intrusive.

I also get a vertical scrollbar
![Screenshot 2023-10-10 at 12 22 10](https://github.com/hashintel/hash/assets/37743469/6e568205-d40f-4fde-9645-9b06af0f319d)

benwerner01 replied:
I've made the overflow `auto` in https://github.com/hashintel/hash/pull/3338/commits/e3f3674a4d2fdd8ee782d0de480840fd923e454d, I can't reproduce the vertical scrollbar. I don't really get why chrome sometimes adds these intrusive scrollbars, and other times uses the native overlaying ones. I'm not sure its worth modifying the styling of this further as I need to rework this once I implement [the updated designs](https://www.figma.com/file/gydVGka9FjNEg9E2STwhi2/HASH?node-id=9717%3A50623&mode=dev) anyway as part of https://linear.app/hash/issue/H-931/display-cover-image-in-user-profile-page

CiaranMn replied:
We spoke about this and figured out it was because you have overlay scrollbars in Chrome and I have the solid gray ones (that shift the layout). Applying `overflow-y: hidden` to the tabs (somewhat hackily) fixes it

CiaranMn commented on apps/hash-frontend/src/pages/[shortname].page/profile-page-header.tsx:
```diff
@@ -70,42 +89,86 @@ export const ProfilePageHeader: FunctionComponent<{
-            )}
+              {isEditable ? (
+                <>
+                  <Button
```
I would remove the background and border on hover as we don't do this for other similar links... I think the text changing colour is sufficient
![Screenshot 2023-10-10 at 12 23 59](https://github.com/hashintel/hash/assets/37743469/e16934ce-bddb-408d-b913-5965894b0df3)

benwerner01 replied:
This button is going to be removed once I implement [the updated designs](https://www.figma.com/file/gydVGka9FjNEg9E2STwhi2/HASH?node-id=9717%3A50623&mode=dev) as part of https://linear.app/hash/issue/H-931/display-cover-image-in-user-profile-page, so I think I'll just leave it as is for now.

CiaranMn replied:
Fair enough!

CiaranMn reviewed (commented):
Thanks Ben, looking good! A few suggestions for code + UX/styling improvements

benwerner01 commented:
> I've got a duplicate entity type appearing in the list a few times. This may just be a dev/staging bug, but we should probably make sure logic for deduplicating the list is sound, and ensuring that all 5 selected entity types are different. I don't think we're showing already-listed entity types in the selector list, as I wasn't able to add the entity type an additional time, but we should also make sure this is the case. Maybe removal/re-adding causes duplication? I haven't tried to debug why, so will leave to you.

@nonparibus I'm not sure how it was possible you got into this state, perhaps it's related to the deployment not having the necessary backend changes, or if you tested the PR before it was ready. Although it should not be possible to select an entity type in the selector that has already been selected, as an additional precaution I've added a filter for duplicates in the update function in https://github.com/hashintel/hash/pull/3338/commits/70dac57bb729013da7e4e858049716e15082302e.

CiaranMn reviewed (approved):
:shipit:
